### PR TITLE
[shaderc] Use build type to build targets

### LINKED ISSUE #16122: vcpkg integrate install logic is incorrect
**Describe the bug**
vcpkg integrate install does not apply the correct logic.  
It automatically adds all libraries under installed/x64-windows/lib as linker arguments
Some packages install multiple versions of libraries.
The user is only supposed to link against one specific version of the library, as specified via a pkg-config file or cmake toolchain file. 
Otherwise we get linker errors about a symbol already defined

**Environment**
- OS: Windows
- Compiler: MSVC 2019

**To Reproduce**
vcpkg install shaderc:x64-windows
vcpkg integrate install

**Expected behavior**
The current logic for vcpkg integrate install is not correct.
VCPKG cannot just add all libraries in installed/x64-windows/lib as linker inputs.
The user should specify the linker inputs either explicitly or via some tool such as pkg-config --libs or via a cmake toolchain file.

**Failure logs** 
untitled:3408: 2>    shaderc.lib(shaderc.cc.obj) : error LNK2005: shaderc_compile_into_spv already defined in shaderc_shared.lib(shaderc_shared.dll)

**Additional context**

### PR DESCRIPTION
**Pull request by JackBoosY**
Currently, shaderc has generated dynamic libraries, static libraries and static libraries containing all dependent binaries at the same time.
This PR will:
- Determine whether to generate a dynamic library or a static library according to `BUILD_SHARED_LIBS`.
- Disable generating static libraries containing all dependent binaries.

Partly solved #15452.
Fixes #16122.

### CODE REVIEW

File: ports/shaderc/portfile.cmake [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
     OUT_SOURCE_PATH SOURCE_PATH
     REPO google/shaderc
     REF 05c766a1f4615ee37d37d09271aaabca30ffd293
     SHA512 329697e8e23d619313440d57ef740a94c49d13533e1b8734fc8ff72fd5092c2addabb306f64cb69160fa5fee373a05ba39a5ee6d92d95e5e2e9c7ec96a51aadc
     HEAD_REF master
     PATCHES 
-    	"disable-update-version.patch"
-        "fix-install.patch"
+    	disable-update-version.patch
+        fix-install.patch
```
Comment (NancyLi1013) [inline ports/shaderc/portfile.cmake:12]:
```suggestion
        disable-update-version.patch
        fix-install.patch
```

File: ports/shaderc/portfile.cmake [hunk 2]
```diff
@@ -31,21 +33,23 @@ vcpkg_add_to_path(PREPEND "${PYTHON3_EXE_PATH}")
...
 
-#Safe to remove as libs are static
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
 file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
 
-# Handle copyright
-file(COPY ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/shaderc)
-file(RENAME ${CURRENT_PACKAGES_DIR}/share/shaderc/LICENSE ${CURRENT_PACKAGES_DIR}/share/shaderc/copyright)
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/shaderc" RENAME copyright)
```
Comment (NancyLi1013) [inline ports/shaderc/portfile.cmake:55]:
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Review by NancyLi1013 — APPROVED** [review]

**Comment by JackBoosY** [comment]
Waiting for merge #16138.

**Comment by ras0219-msft** [comment]
This LGTM

**Comment by jmoguillansky-gpsw** [comment]
Hi @JackBoosY I updated vcpkg to the latest version.  It seems that only static shaderc was built.
How do I build the shared lib?
Am I supposed to set BUILD_SHARED_LIBS environment variable or something?

**Comment by jmoguillansky-gpsw** [comment]
Seems that this PR disables shaderc shared library build?

#Note: glslang and spir tools doesn't export symbol and need to be build as static lib for cmake to work
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

This part doesn't make sense.  If glslang and spir-tools require static shaderc then they can link against the static shaderc, that doesn't mean that shaderc doesn't support shared library build?

**Comment by JackBoosY** [comment]
@jmoguillansky-gpsw For building shared, use command `./vcpkg install shaderc`.
You should use the right triplet to build it.

**Comment by jmoguillansky-gpsw** [comment]
> @jmoguillansky-gpsw For building shared, use command `./vcpkg install shaderc`.
> You should use the right triplet to build it.

Hi @JackBoosY ,
Yes I tried that: vcpkg install shaderc:x64-windows, but it still only built the static lib?

**Comment by JackBoosY** [comment]
@jmoguillansky-gpsw  Ops, sorry, I forgot we force it to build as a static library.
I will investigate later.

**Comment by ras0219-msft** [comment]
The issue is not that glslang and spir-tools require static shaderc -- rather, the problem is that glslang and spir-tools can only be built statically, which means building a shared library for shaderc is _extremely_ suspect (it will contain a duplicate copy of the glslang/spir-tools code along with the main application binary as well as any other consumers). Shaderc consumes glslang, not the other way around.

Therefore, until glslang and spir-tools can be built dynamically, we do not want to build shaderc dynamically.

**Comment by jmoguillansky-gpsw** [comment]
Shaderc upstream supports shared lib build, and its shipped as part of vulkan sdk.  There's no issue with shaderc , the tool just decides to use the static lib, thats all

**Comment by jmoguillansky-gpsw** [comment]
I think there's a broader discussion with regards to vcpkg architecture.  Each module doesn't know how it's used, only what it can build.  Some executables may choose to always use static lib, some shared lib, etc.  From my perspective the vcpkg architecture should accomodate these cases.  
One option:
The module should build all supported configurations (static, shared, etc).  
Vcpkg integrate install can have a param of default lib type (e.g. shared or static) the same way that a linker (ld on linux) has a param to specify default lib type.

**Comment by jmoguillansky-gpsw** [comment]
Regarding vcpkg integrate install:
Maybe one option to consider is:
let the user navigate to a project and type something like "vcpkg integrate <module name>" and then vcpkg will integrate the module into the project.
This gives the user convenience, but perhaps safer than making all libs globally available for all msvc projects,etc.

**Comment by JackBoosY** [comment]
glslang and spir-tools only depend on the static library, not dynamic.

**Pull request merged**